feat(sdk): Use stable IDs in SDK components

oisincoveney opened the pull request:
Uses https://github.com/metabase/metabase/pull/47094 to allow for stable IDs in SDK.

**_How to test_**
Now, you should just be able to use an entity ID in the `InteractiveQuestion` or `InteractiveDashboard` components and it should ✨ just work ✨.

Right now, the best way to get an entity ID is to examine the network request when you are loading a dashboard or a question. When you go to a dashboard or question page, open up the `Network` Devtools tab, find a dashboard/question API call that loads the data, and then find the `entity_id` entry. You can then use that key for testing.

Ideally we'll have a _much_ easier way of getting this key in the future. For now, this is what we have to do 🙃 

<img width="872" alt="image" src="https://github.com/user-attachments/assets/23702025-9fb8-4e06-aeb7-c33389e667a1">

<img width="332" alt="image" src="https://github.com/user-attachments/assets/d3c5e2f1-7057-412d-bda6-42966e23de41">

escherize committed:
entity id translation + tests

escherize committed:
Merge branch 'master' into bcm-entity-id-translation-service

escherize committed:
add api level test

escherize committed:
Merge branch 'bcm-entity-id-translation-service' of github.com:metabase/metabase into bcm-entity-id-translation-service

escherize committed:
simplify definition of eid-table->model + add test

escherize committed:
update tests to take keywords

escherize committed:
improve comment

escherize committed:
generate the eid-table->model map

escherize committed:
delete now-obsolete test

escherize committed:
make it work in oss

escherize committed:
put the resulting response into a key, so we can add more information later

escherize committed:
formatting

escherize committed:
use model names without the model/ prefix as keys

Commit:
Merge branch 'master' into entity-id-translation-service

escherize committed:
Creates list of `api/model->db-model`

- update keys for util/entity_id request
- update shape of util/entity_id response
- add test for not-found eids

escherize committed:
Merge branch 'entity-id-translation-service' of github.com:metabase/metabase into entity-id-translation-service

escherize committed:
formatting

escherize committed:
Merge branch 'master' into entity-id-translation-service

Commit:
Merge branch 'master' into entity-id-translation-service

Commit:
Add API, hook for SDK, and types

escherize committed:
Merge branch 'entity-id-translation-service' of github.com:metabase/metabase into entity-id-translation-service

escherize committed:
Respond to code review feedback

Commit:
Improve hook

Commit:
Move hook to common file

Commit:
Merge branch 'master' into entity-id-translation-service

Commit:
Merge branch 'entity-id-translation-service' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Fix some types

Commit:
Merge branch 'master' into entity-id-translation-service

Commit:
Merge branch 'entity-id-translation-service' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Add id=nil for lost IDs

Commit:
ensure that types are fixed by moving some code around

Commit:
Attempt to fix more types

Commit:
Attempt to fix even more types

Commit:
Finally?

Commit:
Merge branch 'master' into entity-id-translation-service

Commit:
Merge branch 'entity-id-translation-service' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

WiNloSt commented on frontend/src/metabase/lib/entity-id/hooks/use-valid-id.ts:
```diff
@@ -0,0 +1,35 @@
+    data: entity_ids,
+    isError,
+    isLoading,
+  } = useTranslateEntityIdQuery({
```
I think we could be a little bit smarter before calling this endpoint which might save time and improve performance. It should be fairly simple to detect if a given ID is an entity ID, it should be comprised of these allowed strings

https://github.com/metabase/metabase/blob/ebf18f35b22d16bc286b3cda9c29b571a815b714/src/metabase/util/jvm.clj#L28

heypoom replied:
+1, if we can have a basic heuristic that detect if it's a valid NanoID first, it could reduce the overhead of this API call.

oisincoveney replied:
Yep, the only issue is with the stupid rule of hooks in terms of conditionally calling the API. I _think_ I can do something within the API file itself so let me try that first

oisincoveney replied:
ok, looks like there's an escape hatch with the skip flag so I'll use that

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/context/InteractiveQuestionProvider.tsx:
```diff
@@ -89,6 +91,20 @@ export const InteractiveQuestionProvider = ({
+}: Omit<InteractiveQuestionProviderProps, "options"> & {
+  options?: QueryParams;
+}) => {
+  const id = useValidIdForEntity({
```
Is this because we want to preserve `InteractiveQuestionProviderInner` to still work with `cardId` for other usages of this provider?

oisincoveney replied:
nope, this should work with all question components that use InteractiveQuestionProvider. I didn't want to put it in the actual components since we might be using the provider for other question components in the future and I didn't want to have to put the entity ID logic in all of the places

heypoom commented on enterprise/frontend/src/embedding-sdk/components/public/InteractiveDashboard/InteractiveDashboard.tsx:
```diff
@@ -110,6 +112,16 @@ const InteractiveDashboardInner = ({
+export const InteractiveDashboard =
+  withPublicComponentWrapper<InteractiveDashboardProps>(
+    ({ dashboardId, ...rest }) => {
+      const id = useValidIdForEntity({
```
A bit curious about why we don't add `useValidIdForEntity` into the `InteractiveDashboardInner` component itself, if the input `id` would support both stable ids and numerical ids?

oisincoveney replied:
Ah! I think there's some misunderstanding here (most likely caused by me, don't worry). The hook will take in anything - whether that's an entity ID or a numerical ID or whatever, and if the entity ID can't be found then we return the value that was passed in. So basically if we were to pass a numerical ID then the hook would return the numerical ID. 

I'll try to make this clearer in the code so it's more obvious

heypoom commented on enterprise/frontend/src/embedding-sdk/components/public/InteractiveQuestion/InteractiveQuestion.tsx:
```diff
@@ -22,10 +22,10 @@ import {
 
 export type InteractiveQuestionProps = PropsWithChildren<{
-  questionId?: CardId;
+  questionId?: Card["id"];
```
I think `CardId` should point to the same as `Card["id"]`, from `frontend/src/metabase-types/api/card.ts`:

```ts
export interface Card<Q extends DatasetQuery = DatasetQuery>
  extends UnsavedCard<Q> {
  id: CardId;
```

oisincoveney replied:
yep I get it

heypoom commented on enterprise/frontend/src/embedding-sdk/components/public/InteractiveQuestion/InteractiveQuestion.tsx:
```diff
@@ -39,6 +39,10 @@ export const _InteractiveQuestion = ({
   children = null,
 }: InteractiveQuestionProps &
   InteractiveQuestionResultProps): JSX.Element | null => {
+  if (!questionId) {
+    return <div>Loading...</div>;
+  }
```
Should we use the (user-supplied) loader component we have instead?

oisincoveney replied:
yes!

heypoom commented on frontend/src/metabase-types/api/mocks/actions.ts:
```diff
@@ -34,6 +34,7 @@ export const createMockQueryAction = ({
 }: Partial<WritebackQueryAction> = {}): WritebackQueryAction => {
   return {
     id: 1,
+    entity_id: "ACTIONmetabaseACTIONs",
```
Should we extract this to a constant for easy testing of entity ids, similar to `ORDER_ID` we have in e2e helpers? My guess is that this weird format in `ACTIONmetabaseACTIONs ` is the format of entity ids

oisincoveney replied:
yeah that's fair. I was feeling a little silly typings these 😅

oisincoveney replied:
Ok, I made a `createMockEntityId` helper and tightened `BaseEntityId`

heypoom reviewed (approved):
Looks great! Added some questions and notes on my end.

PS. I wonder if numerical IDs are still supported with this change, or only stable IDs are supported?

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Add createMockEntityId and tighten BaseEntityID type to only accept nano IDs

oisincoveney commented:
@heypoom @WiNloSt I've made some changes to the typing for BaseEntityID and added some type guards. I think it'll be better in terms of clamping down on types (i.e. not allowing _any_ string in the `entity_id` field) and added a `createMockEntityId` function so that we don't have random strings all the time.

The commit to view is here: [Add createMockEntityId and tighten BaseEntityID type to only accept n…](https://github.com/metabase/metabase/pull/47210/commits/1f5d6546a3c0f5f998f12b799b2719df7ce65839)

heypoom commented on frontend/src/metabase-types/api/entity-id.ts:
```diff
@@ -0,0 +1,24 @@
+  );
+};
+
+export const asNanoID = (id: string): NanoID => {
```
(note: looks like we don't reference `asNanoID` anywhere yet in this PR)

oisincoveney replied:
with all of these new type errors, I think I'm about to. I'll make those changes first

heypoom commented on frontend/src/metabase/api/entity-id.ts:
```diff
@@ -0,0 +1,85 @@
+        Object.values(requestEntities)
+          .flat()
+          .forEach(entityId => {
+            if (isNanoID(entityId)) {
+              throw new Error(`Entity ID ${entityId} must be a NanoID`);
+            }
```
Should this be `if (!isNanoID(entityId))` instead?

oisincoveney replied:
woah, you reviewed too quickly 😂

heypoom replied:
whoops, my bad 😆

oisincoveney replied:
issue is resolved 😄

heypoom reviewed (changes requested):
I think the `isNanoID` check should be negated - we should only throw an error when the entity id is not a nanoId

Commit:
Rename file

heypoom reviewed (approved):
Changes looks good! Looks like we're doing the entity ID validation in the `entityIdApi` itself, which is nice.

Commit:
Fix type errors on tabS

WiNloSt commented on frontend/src/metabase-types/api/mocks/entity-id.ts:
```diff
@@ -0,0 +1,5 @@
+import type { BaseEntityId } from "../entity-id";
+
+export const createMockEntityId = (): BaseEntityId => {
+  return "abcdefghijklmnopqrstu" as BaseEntityId;
```
nit: I'd expect this function to return something unique. This might not be apparent now, but when we need to have different entity_id for reference this could make tests behave unexpectedly. I think it's fine for now if we don't want to do anything about it.

But can you at least have a todo comment that this needs to either:
1. Be implemented or
2. Use the ID generation function from BE (which isn't exported currently)

oisincoveney replied:
I'll just create a simple generator right now and add a TODO. Maybe we can move this to CLJS in the future

oisincoveney replied:
ah! we can use the `nanoid` package supplied by rtk for now 😄 much better

WiNloSt commented on frontend/src/metabase-types/api/entity-id.ts:
```diff
@@ -0,0 +1,24 @@
+const NANOID_ALPHABET =
+  "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
```
```suggestion
  "_-0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
```

This is taken from our Clojure implementation.

https://github.com/metabase/metabase/blob/44888103944f545c3b6c84c489832fdc2868e534/src/metabase/util/jvm.clj#L28

Commit:
hopefully fix backend tests

WiNloSt commented on frontend/src/metabase-types/api/entity-id.ts:
```diff
@@ -0,0 +1,24 @@
+export const isNanoID = (id: unknown): id is NanoID => {
+  return (
+    typeof id === "string" &&
+    id.length === NANOID_LENGTH &&
+    new RegExp(`^[${NANOID_ALPHABET}]+$`).test(id)
```
you could combine these lines to
```suggestion
    new RegExp(`^[${NANOID_ALPHABET}]{${NANOID_LENGTH}}$`).test(id)
```

Idk if the above syntax is actually correct or not, but I want to result to be something like `/^[_-abc...]{21}$/`

oisincoveney replied:
Is that necessarily better? I kinda like the obvious explicit check to the length - it's easier to read

WiNloSt replied:
not necessarily better I'd say. It's just that I'm so used to use Reg ex to also check the string length, so I thought that might be a bit more streamline, but in this case the number isn't hard-coded, and using a variable makes the regex construction looks too complicated. Feels free to leave it as is.

WiNloSt commented on frontend/src/metabase-types/api/entity-id.ts:
```diff
@@ -0,0 +1,24 @@
+
+declare const __brand: unique symbol;
+type Brand<T, B> = T & { readonly [__brand]: B };
+type NanoID = Brand<string, "NanoID">;
```
2 questions:
1. How does this type work?
    I tried `const: NanoID = "abcdefghijklmnopqrstu"` and this gives me error. I'd expect this to work given it has a type.
2. We export `NanoID` as `BaseEntityId` while the function is named `isNanoID` I think we should use the same name either a `Nano ID` or an `Entity ID`

oisincoveney replied:
1. The idea is that you can't just create a NanoID - those should really be always coming from the backend. In your case, you've created a `string` that you're trying to specify as a `NanoID`. To make this work, you'll need to explicitly cast the string to a NanoID with `as NanoID/BaseEntityId` or use the `asNanoId` utility below to keep type safety.
2. As I was typing that first answer I see what you mean. I'll stick with `EntityId` for uses outside this file because it's super annoying otherwise

WiNloSt replied:
Re 1) that makes sense. This ID isn't supposed to be manually created.

oisincoveney replied:
I'll add a comment at the top of this file just to clarify for anybody else who might have the same idea. It's a very valid way of thinking

Commit:
Add generator function for mockEntityId and add comment to entity-id file

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Add generator function for mockEntityId and add comment to entity-id file

WiNloSt commented on frontend/src/metabase-types/api/entity-id.ts:
```diff
@@ -0,0 +1,42 @@
+export const NANOID_LENGTH = 21;
+
+declare const __brand: unique symbol;
+type Brand<T, B> = T & { readonly [__brand]: B };
```
What is `__brand` for? is it just a random symbol to make sure a string couldn't match this type?

WiNloSt commented on frontend/src/metabase-types/api/entity-id.ts:
```diff
@@ -0,0 +1,42 @@
+  );
+};
+
+export const asBaseEntityID = (id: string): BaseEntityId => {
```
just another name nit. What's the difference between a Base entity ID and an Entity ID?

When there's a Base something, does that mean there would be its counterpart Something?

WiNloSt commented on frontend/src/metabase-types/api/dashboard.ts:
```diff
@@ -167,11 +169,11 @@ export type DashboardTabId = number;
   position?: number;
-  created_at: string;
-  updated_at: string;
+  created_at?: string;
+  updated_at?: string;
```
why do we need to change the type here?

oisincoveney replied:
They're not populated anywhere in the frontend - to create a new tab, we use `getDefaultTab` which initializes the three properties there to empty strings, but they're not manipulated and nothing is done to them. We should just leave them as things that will be populated by the server once we save the dashboard tabs

WiNloSt commented on enterprise/frontend/src/embedding-sdk/components/public/InteractiveDashboard/InteractiveDashboard.tsx:
```diff
@@ -110,6 +112,16 @@ const InteractiveDashboardInner = ({
+  withPublicComponentWrapper<InteractiveDashboardProps>(
+    ({ dashboardId, ...rest }) => {
+      const id = useValidatedEntityId({
+        type: "dashboard",
```
There is one DX improvement we could do here (I'm not sure how complicated this would be tho), since we know the type and possible values are finite. We could probably infer the return type automatically.

Imagine passing `type: "dashboard"` and the hook returning `DashboardId`, that'd be AWESOMEEE.

One pro if we pulled that out is to avoid human error since the caller could practically cast the return value to any type.

oisincoveney replied:
I took like 3-4 hours trying to figure that one out and I can't seem to get it to work 🙃  basically, after creating the mapped types + generics, the TS compiler would just return `any`. I'm really sad about it

WiNloSt replied:
No worries then. I don't think it'd be that simple 😓

WiNloSt reviewed (approved):
Looks good to me.

Commit:
Fix unit tests, hopefully for real this time

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Rig the type to work

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Fix readme and validated entity IDs for static and interactive questions

Commit:
Merge remote-tracking branch 'origin/feat(sdk)/use-stable-ids-in-sdk-components' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

npretto committed:
some ts magic to make everybody happy 🪄

npretto commented on frontend/src/metabase/lib/entity-id/hooks/use-validated-entity-id.ts:
```diff
@@ -0,0 +1,77 @@
+ *
+ */
+export const useValidatedEntityId = <
+  TEntity extends keyof SUPPORTED_ENTITIES,
```
Here's the result of this TS unreadable code
https://github.com/user-attachments/assets/33ad705a-f954-4ade-be39-6928a60de5a9

npretto commented on frontend/src/metabase/lib/entity-id/hooks/use-validated-entity-id.ts:
```diff
@@ -0,0 +1,77 @@
+ * @param {string | number | null | undefined} params.id - The ID to validate and potentially translate.
+ *
+ */
+export const useValidatedEntityId = <
```
This names makes me think this hook an id and returns a "validated entity id", I know it doesn't sound pretty but I'd call it something like `useMaybeTranslateEntityId` since that's kinda what it does, it **maybe** translated an entity id to an id

npretto commented on frontend/src/metabase/lib/entity-id/hooks/use-validated-entity-id.ts:
```diff
@@ -0,0 +1,77 @@
+  type: TEntity;
+  id: BaseEntityId | string | number | null | undefined;
+}):
+  | { id: TReturnedId; isLoading: false; isError: false }
```
I think it's possible to make this more generic, but I tried and the code will be even more readable

npretto commented:
I talked with Oisin and he let me push some TS magic to make the hook easier to use and with a stricter typing: c470b4dfd9af2bd4595dd613e9981593f6f1c70a

npretto committed:
put back use memo

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
A wild symbol appeared!

Commit:
Revert status change to 'ok' from 'success'

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components

Commit:
Merge branch 'master' into feat(sdk)/use-stable-ids-in-sdk-components
